lib/internal/worker.js

```diff
@@ -0,0 +1,475 @@
+'use strict';
+
+const EventEmitter = require('events');
+const assert = require('assert');
+const path = require('path');
+const util = require('util');
+const { Readable, Writable } = require('stream');
+const {
+  ERR_INVALID_ARG_TYPE,
+  ERR_WORKER_NEED_ABSOLUTE_PATH,
+  ERR_WORKER_UNSERIALIZABLE_ERROR,
+  ERR_WORKER_UNSUPPORTED_EXTENSION,
+} = require('internal/errors').codes;
+
+const { internalBinding } = require('internal/bootstrap/loaders');
+const { MessagePort, MessageChannel } = internalBinding('messaging');
+const { clearAsyncIdStack } = require('internal/async_hooks');
+const { serializeError, deserializeError } = require('internal/error-serdes');
+
+util.inherits(MessagePort, EventEmitter);
+
+const {
+  Worker: WorkerImpl,
+  getEnvMessagePort,
+  threadId
+} = internalBinding('worker');
+
+const isMainThread = threadId === 0;
+
+const kOnMessageListener = Symbol('kOnMessageListener');
+const kHandle = Symbol('kHandle');
+const kName = Symbol('kName');
+const kPort = Symbol('kPort');
+const kPublicPort = Symbol('kPublicPort');
+const kDispose = Symbol('kDispose');
+const kOnExit = Symbol('kOnExit');
+const kOnMessage = Symbol('kOnMessage');
+const kOnCouldNotSerializeErr = Symbol('kOnCouldNotSerializeErr');
+const kOnErrorMessage = Symbol('kOnErrorMessage');
+const kParentSideStdio = Symbol('kParentSideStdio');
+const kWritableCallbacks = Symbol('kWritableCallbacks');
+const kStdioWantsMoreDataCallback = Symbol('kStdioWantsMoreDataCallback');
+const kStartedReading = Symbol('kStartedReading');
+const kWaitingStreams = Symbol('kWaitingStreams');
+const kIncrementsPortRef = Symbol('kIncrementsPortRef');
+
+const debug = util.debuglog('worker');
+
+// A communication channel consisting of a handle (that wraps around an
+// uv_async_t) which can receive information from other threads and emits
+// .onmessage events, and a function used for sending data to a MessagePort
+// in some other thread.
+MessagePort.prototype[kOnMessageListener] = function onmessage(payload) {
```

Comment:
hmmm... I can't say that I'm too fond of mutating the `MessagePort` like this then exposing it to users, especially with an enumerable `_`-prefixed property (attached below). Most other places we have a JavaScript object wrapper with an internal `handle`. Is there a particular reason you went with this model, @addaleax?

Reply:
> Is there a particular reason you went with this model

Yes: `MessagePort`s are themselves transferrables in the context of `postMessage()`, but as far as the V8 serialization API is concerned, only host objects (i.e. C++-backed) can be transferrables. I think that makes sense from a serializer’s point of view.

If we were to wrap it in a JS object, the serialization algorithm would loose its prototype information and it would come out as a plain JS object with a `_handle` property, rather than our JS wrapper, on the other side.

> especially with an enumerable `_`-prefixed property (attached below)

Right … so … I agree. We shouldn’t be doing that. It should be a symbol but we don’t currently have a good story around those on the native side. I added 8e0442cbda96 and da23c3f021d8 for that to this PR, please take a look at it.

Reply:
Definitely not blocking for now. I'll give some thought to alternatives tho :-)